tests/unittests/sources/azure/test_imds.py

```diff
@@ -481,6 +481,46 @@ def test_non_json_repsonse(
             )
         ]
 
+    def test_logs_unique_errors(
+        self,
+        caplog,
+        mock_requests_session_request,
+        mock_url_helper_time_sleep,
+    ):
+        fake_md = {"foo": {"bar": []}}
+        fake_errors = [
+            fake_http_error_for_code(404),
+            fake_http_error_for_code(410),
+            fake_http_error_for_code(429),
+            fake_http_error_for_code(500),
+            requests.ConnectionError("Fake connection error"),
+            requests.Timeout("Fake connection timeout"),
+        ]
+        attempts = len(fake_errors) + 1
+        mock_requests_session_request.side_effect = fake_errors + [
+            mock.Mock(content=json.dumps(fake_md))
+        ]
+
+        md = imds.fetch_metadata_with_api_fallback(retry_deadline=300)
```

Comment:
This test case doesn't exercise the new code when `log = False` we never fall into a case here where [`self._request_count < self._logging_threshold`](https://github.com/canonical/cloud-init/pull/4468/files#diff-23c0349c65e82f104b4f4df89065e282abd73e7d64a07098383d8b76c7755bc4R82) which means we log is always True as all logs are persisted.


You can see this by removing the changset from imds.py and re-running the test to success or by adding duplicate fake_errors and still seeing an error logged for every duplicate index.


  I don't think `fetch_metadata_with_api_fallback` gives us a chance to provide a ReadUrlRetryHandler with a logging_backoff of the default 1.0.  Wouldn't we need to use `fetch_reprovision_data` function instead as it sets a logging backoff which will give us a chance to set `log = False` in handling.

This test should also attempt to provide duplicated error codes or error types to ensure that you don't get duplicate logs when you are below the next logging_threshold but still have duplicated errors reported.